File: 02.実践帳票サンプル集－様々な帳票手法/VB/WinForm/Form1.vb
﻿Imports Pao.Reports
Imports System.IO

Namespace Sample
    Partial Public Class Form1
        Inherits Form


        Private Sub Form1_Load(sender As Object, e As EventArgs) Handles MyBase.Load
            Me.Top = 5
            'Me.Height = Screen.PrimaryScreen.WorkingArea.Height - 50

            ' VB.NET との共有リソースパス(Util.SharePath)設定
            Util.SetSharePath()

            ' サンプル帳票初期値セット
            cmbReportType.SelectedIndex = 0
        End Sub


        Private dataLoaded As Boolean = False

        Private Sub btnExecute_Click(sender As Object, e As EventArgs) Handles btnExecute.Click
            'IReport インターフェースで宣言(印刷・レポートどちらでも使える入れ物の用意)
            Dim paoRep As IReport

            '■インスタンスの生成がエラーとなる場合の対処法下記コメントを外してください
            'System.Text.Encoding.RegisterProvider(System.Text.CodePagesEncodingProvider.Instance)

            ' *** 出力別にインスタンス生成 ***

            If radPreview.Checked Then ' WPFプレビュー が選択されている場合
                'WPF版プレビューオブジェクトのインスタンスを獲得
                paoRep = ReportCreator.GetPreviewWpf()
            ElseIf radPrint.Checked OrElse radXPS.Checked Then ' 印刷、又は、XPS出力 が選択されている場合
                '印刷オブジェクトのインスタンスを獲得
                paoRep = ReportCreator.GetReport()
            ElseIf radPDF.Checked Then ' PDFが選択されている場合
                If cmbReportType.Text.Contains("広告") Then
                    'イメージPDF出力オブジェクトのインスタンスを獲得
                    paoRep = ReportCreator.GetImagePdf()
                Else
                    'PDF出力オブジェクトのインスタンスを獲得
                    paoRep = ReportCreator.GetPdf()
                End If
            Else
                '印刷オブジェクトのインスタンスを獲得
                paoRep = ReportCreator.GetReport()
            End If

            ' **** 帳票別に帳票データをセット ***

            If cmbReportType.Text.Contains("郵便番号一覧") Then
                Make郵便番号一覧.SetupData(paoRep)
            ElseIf cmbReportType.Text.Contains("見積書") Then
                Make見積書.SetupData(paoRep)
            ElseIf cmbReportType.Text.Contains("請求書") Then
                Make請求書.SetupData(paoRep)
            ElseIf cmbReportType.Text.Contains("商品大小分類") Then
                Make商品大小分類.SetupData(paoRep)
            ElseIf cmbReportType.Text.Contains("デザイン変更") Then
                Dim defIndex As Integer = 0
                If radDesign2.Checked Then defIndex = 1

                If dataLoaded Then
                    ' デザインファイルのみ変更
                    paoRep.ChangeDefFile(Util.GetDefFile(defIndex))
                Else
                    Makeデザイン変更.SetupData(paoRep, defIndex)
                    dataLoaded = True
                End If
            ElseIf cmbReportType.Text.Contains("広告") Then
                Make広告.SetupData(paoRep)
            End If

            ' データ読み込み済みフラグをクリアしておかないと
            If Not cmbReportType.Text.Contains("デザイン変更") Then dataLoaded = False

            ' *** 各種出力 ***

            If radPreview.Checked OrElse radPrint.Checked Then '印刷・プレビューが選択されている場合
                paoRep.Output() ' 印刷又はプレビューを実行
            ElseIf radPDF.Checked Then 'PDF出力が選択されている場合
                Dim saveFileName As String = ShowSaveDialog("pdf") ' PDFファイル保存ダイアログ
                If saveFileName = "" Then Return

                'PDFの保存
                paoRep.SavePDF(saveFileName)

                OpenSaveFile(saveFileName) ' 保存したPDFファイルを開く
            ElseIf radSVG.Checked Then 'SVG出力が選択されている場合
                Dim saveFileName As String = ShowSaveDialog("html") ' SVGファイル保存ダイアログ
                If saveFileName = "" Then Return

                'SVGデータの保存
                paoRep.SaveSVGFile(saveFileName)

                OpenSaveFile(saveFileName) ' 保存したSVGファイルを開く
            ElseIf radXPS.Checked Then 'XPS出力が選択されている場合
                Dim saveFileName As String = ShowSaveDialog("xps") ' XPSファイル保存ダイアログ
                If saveFileName = "" Then Return

                'XPS印刷データの保存
                paoRep.SaveXPS(saveFileName)

                OpenSaveFile(saveFileName) ' 保存したXPSファイルを開く
            End If
        End Sub

        '*** 共通処理をメソッド化

        ' 指定された種類のファイル保存ダイアログを開き
        ' 確定した場合保存ファイル名(フルパス)を返す。確定しない場合空文字を返す。
        Private Function ShowSaveDialog(type As String) As String
            Dim dlg As New SaveFileDialog()
            dlg.FileName = "印刷データ"
            dlg.DefaultExt = "." & type.ToLower()
            dlg.Filter = type.ToUpper() & " Document (." & type.ToLower() & ")|*." & type.ToLower() ' Filter files by extension
            dlg.InitialDirectory = Environment.GetFolderPath(Environment.SpecialFolder.Desktop)

            ' Show save file dialog box
            Dim result As DialogResult = dlg.ShowDialog()

            ' Process save file dialog box results
            If result = DialogResult.OK Then
                Return dlg.FileName
            End If
            Return ""
        End Function

        ' 保存したファイルの起動
        Private Sub OpenSaveFile(filePath As String)
            Dim type As String = Path.GetExtension(filePath)?.TrimStart("."c).ToUpperInvariant()

            If MessageBox.Show("保存した " & type & " を表示しますか？", type & " の表示", MessageBoxButtons.YesNo) = DialogResult.Yes Then
                Dim startInfo As New System.Diagnostics.ProcessStartInfo(filePath)
                startInfo.UseShellExecute = True
                System.Diagnostics.Process.Start(startInfo)
            End If
        End Sub

        Private Sub cmbReportType_SelectedIndexChanged(sender As Object, e As EventArgs) Handles cmbReportType.SelectedIndexChanged
            If cmbReportType.Text.Contains("デザイン変更") Then
                pnl.Top = grpDesign.Top + grpDesign.Height + 4
            Else
                pnl.Top = cmbReportType.Top + cmbReportType.Height + 4
            End If
        End Sub

        Private Sub richTextBox2_LinkClicked(sender As Object, e As LinkClickedEventArgs) Handles richTextBox2.LinkClicked
            Dim psi As New ProcessStartInfo With {
                .FileName = e.LinkText,
                .UseShellExecute = True
            }
            Process.Start(psi)
        End Sub
    End Class
End Namespace

File: 04.帳票作成WEB API－各種クラウドに配置・連携中/CS/Client/WinForm/Form1.cs
using Pao.Reports;
using RestSharp;
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Net;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Net.Http.Json;
using System.Text;
using System.Text.Json;
using System.Text.Json.Serialization;
using System.Threading.Tasks;
using System.Windows.Forms;


namespace WinFormsApp1
{
    public partial class Form1 : Form
    {

        public class PrintData
        {
            [JsonPropertyName("id")]
            public int Id { get; set; }

            [JsonPropertyName("data")]
            public string Data { get; set; }
        }

        public Form1()
        {
            InitializeComponent();
        }

        private string GetWebApiUrl()
        {
            // 接続先URL : デフォルトはローカルでデバッグ環境
            string url = "http://localhost:5555/Pao";

            // ※Azure Win / Azure Linux / AWS EC2 / GCP GCE 外部クラウドのURLは、
            // 変更になる可能性があります。
            // ※接続できないときは、次のサイトに各最新のURLを記載しておきますので
            // http://www.pao.ac/reports.net/cloud-url/
            // ※下の各クラウドのURLを上記サイト内に記載れているURLに書き換えてお試しください。

            if (radWebApiAzureWin.Checked)
            {
                // Azure Windowsサーバに配置してある WEB API を使用
                url = "https://are-are.azurewebsites.net/Pao";
            }
            else if (radWebApiAzureLinux.Checked)
            {
                // Azure Linuxサーバ上に配置してある WEB API を使用
                url = "https://hoi-hoi.azurewebsites.net/Pao";
            }
            else if (radWebApiAws.Checked)
            {
                // AWS EC2 Amazon Lnux 2サーバ上に配置してある WEB API を使用
                //url = "http://ec2-43-207-4-252.ap-northeast-1.compute.amazonaws.com:5001/Pao";

                //AWS EC2 は、URLが変更になるため 弊社サイト上のJSON から現在のURLを取得
                url = ResolveAwsUrl();
            }
            else if (radWebApiGcp.Checked)
            {
                // GCP GCE Debian GNU/Linuxサーバ上に配置してある WEB API を使用
                url = "http://34.145.123.177:5000/Pao";
            }

            return url;
        }

        private static readonly HttpClient _hc = new HttpClient();
        private static string _awsUrlCache; // キャッシュ
        private string ResolveAwsUrl()
        {
            if (!string.IsNullOrEmpty(_awsUrlCache)) return _awsUrlCache;

            const string jsonUrl = "https://www.pao.ac/reports.net/redirect-azure-aws-gcp/aws_api.json";
            const string fallback = "http://ec2-43-207-4-252.ap-northeast-1.compute.amazonaws.com:5001/Pao";

            // aws_api.json内は、こういう形式
            // {"api":"http://ec2-43-207-4-252.ap-northeast-1.compute.amazonaws.com:5001/Pao"}


            try
            {
                // 同期的に取得（UIスレッドデッドロック回避用に ConfigureAwait(false)）
                var json = _hc.GetStringAsync(jsonUrl).ConfigureAwait(false).GetAwaiter().GetResult();

                using (var doc = JsonDocument.Parse(json))
                {
                    if (doc.RootElement.TryGetProperty("api", out var apiElem))
                    {
                        var api = apiElem.GetString();
                        if (!string.IsNullOrWhiteSpace(api))
                        {
                            _awsUrlCache = api;
                            return api;
                        }
                    }
                }
            }
            catch { /* 無視してフォールバック */ }

            return fallback;
        }


        /// <summary>
        /// GETクリック
        /// </summary>
        /// <param name="sender"></param>
        /// <param name="e"></param>
        private async void BtnGet_Click(object sender, EventArgs e)
        {

            // WEB API (REST API) の URL 取得
            string webApiUrl = GetWebApiUrl();


            // WEB API  (REST API) のURLセット
            HttpClient Client = new()
            {
                BaseAddress = new Uri(webApiUrl)
            };
            Client.DefaultRequestHeaders.Accept.Clear();
            Client.DefaultRequestHeaders.Accept.Add(
                new MediaTypeWithQualityHeaderValue("application/json"));

            int id = 0;
            if (opt単純な印刷データ.Checked) id = 0;
            if (opt10の倍数.Checked) id = 1;
            if (opt住所一覧.Checked) id = 2;
            if (opt見積書.Checked) id = 3;
            if (opt請求書.Checked) id = 4;
            if (opt商品一覧.Checked) id = 5;
            if (opt広告.Checked) id = 6;

            // パラメータに Json形式でIDをセットして、帳票データをPrinntData型で取得
            PrintData pd = await Client.GetFromJsonAsync<PrintData>("Pao?Id=" + id.ToString());

            byte[] data = Convert.FromBase64String(pd.Data);

            Output帳票(data);

        }

        /// <summary>
        /// POSTクリック
        /// </summary>
        /// <param name="sender"></param>
        /// <param name="e"></param>
        private void BtnPost_Click(object sender, EventArgs e)
        {

            // WEB API (REST API) の URL 取得
            string webApiUrl = GetWebApiUrl();

            // WEB API  (REST API) のURLセット
            RestClient client = new(webApiUrl);

            RestRequest request = new(Method.POST)
            {
                RequestFormat = DataFormat.Json
            };

            PrintData pdIn = new();

            if (opt単純な印刷データ.Checked) pdIn.Id = 0;
            if (opt10の倍数.Checked) pdIn.Id = 1;
            if (opt住所一覧.Checked) pdIn.Id = 2;
            if (opt見積書.Checked) pdIn.Id = 3;
            if (opt請求書.Checked) pdIn.Id = 4;
            if (opt商品一覧.Checked) pdIn.Id = 5;
            if (opt広告.Checked) pdIn.Id = 6;

            // パラメータに PrintData型をJsonシリアライズしてセット
            request.AddParameter("application/json"
                , JsonSerializer.Serialize(pdIn)
                , ParameterType.RequestBody);

            IRestResponse response = client.Execute(request);

            // Json形式で戻ってきた帳票データを PrintData型にデシリアライズ
            PrintData pdOut = JsonSerializer.Deserialize<PrintData>(response.Content);

            byte[] data = Convert.FromBase64String(pdOut.Data);


            Output帳票(data);

        }

        private void Output帳票(byte[] data)
        {

#if NET5_0 || NET6_0 || NET7_0
            System.Text.Encoding.RegisterProvider(System.Text.CodePagesEncodingProvider.Instance);
#endif

            //IReport インターフェースで宣言(印刷・レポートどちらでも使える入れ物の用意)
            IReport paoRep = null;

            if (radPreview.Checked) //ラジオボタンでプレビューが選択されている場合
            {
                //プレビューオブジェクトのインスタンスを獲得
                paoRep = ReportCreator.GetPreview();
            }
            else if (radPrint.Checked) // 印刷が選択されている場合
            {
                //印刷オブジェクトのインスタンスを獲得
                paoRep = ReportCreator.GetReport();
            }
            else if (radPDF.Checked) // PDFが選択されている場合
            {
                if (!opt広告.Checked)
                {
                    //PDF出力オブジェクトのインスタンスを獲得
                    paoRep = ReportCreator.GetPdf();
                }
                else
                {
                    //イメージPDF出力オブジェクトのインスタンスを獲得
                    paoRep = ReportCreator.GetImagePdf();
                }
            }
            else
            {
                //印刷オブジェクトのインスタンスを獲得
                paoRep = ReportCreator.GetReport();
            }


            paoRep.LoadData(data); //印刷データを読み込む


            if (!radPDF.Checked) //PDF出力が選択されていない場合
            {
                paoRep.Output(); //印刷または、プレビューを実行

            }
            else
            {

                saveFileDialog.FileName = "印刷データ";
                saveFileDialog.DefaultExt = ".pdf";
                saveFileDialog.Filter = "PDF Document (.pdf)|*.pdf"; // Filter files by extension
                saveFileDialog.InitialDirectory = Environment.GetFolderPath(Environment.SpecialFolder.Desktop);

                // Show save file dialog box
                if (saveFileDialog.ShowDialog() != DialogResult.OK) return;


                //PDFの保存
                paoRep.SavePDF(saveFileDialog.FileName);

                if (MessageBox.Show("保存した PDF を表示しますか？", "PDF の表示", MessageBoxButtons.YesNo) == DialogResult.Yes)
                {
                    var startInfo = new System.Diagnostics.ProcessStartInfo(saveFileDialog.FileName)
                    {
                        UseShellExecute = true
                    };
                    System.Diagnostics.Process.Start(startInfo);
                }


            }

        }

    }
}
